### PR TITLE
[task-user-reassign] Assign new owner to ScratchOrg when task reassigned

### PR DESCRIPTION
**Pull request by wlonk**
owl sit

### CODE REVIEW

**Comment by wlonk** [comment]
[When re-assigning a task user role, transfer ownership of existing scratch org instead of deleting it.](https://trello.com/c/hxJDjFuD/136-when-re-assigning-a-task-user-role-transfer-ownership-of-existing-scratch-org-instead-of-deleting-it)

**Comment by wlonk** [comment]
@jgerigmeyer I think this does what it needs to. Is there anything I'm missing?

File: metashare/api/models.py [hunk 1]
```diff
@@ -126,6 +127,13 @@ def instance_url(self):
         except (AttributeError, KeyError):
             return None
 
+    @property
+    def gh_username(self):
+        try:
+            return self.github_account.extra_data["login"]
+        except (AttributeError, KeyError):
+            return None
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metashare/api/models.py]: No, there's not. I hadn't thought that `user.username` would reliably be the same, but it will be. However, I suppose there's the concern that someone changes their GitHub username, but in that case many things would get out of sync, and that's a larger issue.

File: metashare/api/models.py [hunk 2]
```diff
@@ -425,6 +433,41 @@ class Task(
...
+
+        super().save(*args, **kwargs)
+
+    def set_owner_values(self):
+        # Dev:
+        new_matching_sa = (
+            self.assigned_dev
+            and SocialAccount.objects.filter(
+                provider="github", uid=str(self.assigned_dev["id"]),
+            ).first()
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metashare/api/models.py]: We can do that; given that I thought I was getting data off of a `SocialAccount`'s `extra_data`, I did it this way. Given that, as you point out, the data is on the `User`, I can do that.

File: metashare/api/models.py [hunk 3]
```diff
@@ -425,6 +433,41 @@ class Task(
...
+        )
+        if new_matching_sa:
+            new_matching_user = new_matching_sa.user
+            ScratchOrg.objects.filter(task=self, org_type=SCRATCH_ORG_TYPES.Dev).update(
+                owner=new_matching_user,
+                owner_sf_username=new_matching_user.sf_username,
+                owner_gh_username=new_matching_user.gh_username,
+            )
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metashare/api/models.py]: I think I like that. So like this flow: https://drive.google.com/file/d/1p4V4Im_doBVaEIzw4gvkj_L4DQox4V1p/view?usp=sharing

File: metashare/api/models.py [hunk 4]
```diff
@@ -425,6 +433,41 @@ class Task(
...
+            and SocialAccount.objects.filter(
+                provider="github", uid=str(self.assigned_qa["id"]),
+            ).first()
+        )
+        if new_matching_sa:
+            new_matching_user = new_matching_sa.user
+            ScratchOrg.objects.filter(task=self, org_type=SCRATCH_ORG_TYPES.QA).update(
+                owner=new_matching_user,
+                owner_sf_username=new_matching_user.sf_username,
+                owner_gh_username=new_matching_user.gh_username,
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metashare/api/models.py]: Those all sound correct, yeah.

[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply metashare/api/models.py]: David confirms: We would need to update the `AdminEmail` in the SF org as well. Question: How to handle scratch org limits for the creating user? Placing this on-hold for now.

File: metecho/api/serializers.py [hunk 5]
```diff
@@ -356,15 +357,46 @@ def update(self, instance, validated_data):
...
+        )
+        new_user = self.get_matching_assigned_user(
+            type_, {f"assigned_{type_}": new_assignee}
+        )
+        if old_user and new_user and old_user.sf_username == new_user.sf_username:
+            return new_user
+        return None
+
+    def get_matching_assigned_user(self, type_, validated_data):
+        assigned = validated_data.get(f"assigned_{type_}", {})
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:395]: Given how this is used, it seems overly complex. Why couldn't this fn just accept `assigned` directly from `_valid_reassign` (either `old_assignee` or `new_assignee`)? That is, why create the `{f"assigned_{type_}": old_assignee}` dict in `_valid_reassign` just to unwrap it here in `get_matching_assigned_user`?

File: metecho/api/serializers.py [hunk 6]
```diff
@@ -356,15 +357,46 @@ def update(self, instance, validated_data):
...
 
+    def _valid_reassign(self, type_, old_assignee, new_assignee):
+        old_user = self.get_matching_assigned_user(
+            type_, {f"assigned_{type_}": old_assignee}
+        )
+        new_user = self.get_matching_assigned_user(
+            type_, {f"assigned_{type_}": new_assignee}
+        )
+        if old_user and new_user and old_user.sf_username == new_user.sf_username:
+            return new_user
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:391]: This looks good to me for reassigning... The next step (re-using a soft-deleted org) will likely require storing either the creating `user` or the creating `user.sf_username` (likely the latter) on the org itself.

File: metecho/api/serializers.py (hunk 5 above)
Comment (wlonk) [reply metecho/api/serializers.py:405]: This was taken from another PR where I had reason to write the logic this way; I wanted it to be compatible with that logic once both are merged.

File: metecho/api/serializers.py [hunk 7]
```diff
@@ -356,15 +357,46 @@ def update(self, instance, validated_data):
                 org_type=SCRATCH_ORG_TYPES.Dev
             )
             for org in orgs:
-                org.queue_delete(originating_user_id=originating_user_id)
+                if new_user := self._valid_reassign(
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metecho/api/serializers.py]: Alas it got complexified out!

File: metecho/api/serializers.py (hunk 6 above)
Comment (wlonk) [reply metecho/api/serializers.py:401]: Can you expand on this comment for me? I'm not sure I follow right now. Thanks!

Comment (jgerigmeyer) [reply metecho/api/serializers.py:401]: @wlonk This PR so far only addresses re-assigning a task-role-with-scratch-org from one user to another. The other part of this story is re-using a soft-deleted scratch org when assigning a task-role to a new user (from no-user-assigned state). To do that, I think we likely want to track the creating `sf_username` on the org itself. See the description on https://trello.com/c/hxJDjFuD/136-when-re-assigning-a-task-user-role-transfer-ownership-of-existing-scratch-org-instead-of-deleting-it

Does that makes sense? I'm happy to jump on a call to talk through it.

Comment (wlonk) [reply metecho/api/serializers.py]: OK, I think that what I have added will do this. Let me know if it works for you, and if it is good, I will sort out the `queue_delete` method, too.

File: metecho/api/models.py [hunk 8]
```diff
@@ -877,19 +878,9 @@ def get_serialized_representation(self, user):
...
-        # flow run.
-        if self.last_modified_at:
-            self.delete_queued_at = timezone.now()
-            self.save()
-            self.notify_changed(originating_user_id=originating_user_id)
-
-        delete_scratch_org_job.delay(self, originating_user_id=originating_user_id)
+        # Formerly a background job, currently synchronous:
+        self.delete()
+        self.finalize_delete(originating_user_id=originating_user_id)
```
Comment (jgerigmeyer) [inline metecho/api/models.py:883]: @wlonk I know you're planning to do more cleanup here. Some of my thoughts:

- When reassigning a role, or removing a user, we're now soft-deleting locally and not deleting on SF.
- But, I think we still want to queue and delete on SF (in a background job) when a user explicitly deletes an org? So I don't think we want to just replace our prior logic for org-deletion -- we're supplementing it.
- Also, before re-assigning and un-deleting an org, we should be sure to check that the org is not expired, or at least handle that case gracefully.

File: metecho/api/models.py [hunk 9]
```diff
@@ -1049,6 +1040,31 @@ def finalize_refresh_org(self, *, error=None, originating_user_id):
             )
             self.queue_delete(originating_user_id=originating_user_id)
 
+    def queue_reassign(self, *, new_user, originating_user_id):
+        from .jobs import user_reassign_job
+
+        self.currently_refreshing_org = True
+        self.deleted_at = None
```
Comment (jgerigmeyer) [inline metecho/api/models.py:1047]: I think this change should maybe happen in `finalize_`, after it's been successfully re-assigned?

File: metecho/api/models.py [hunk 10]
```diff
@@ -877,19 +878,9 @@ def get_serialized_representation(self, user):
...
-        # should only happen when it is destroyed during the initial
-        # flow run.
-        if self.last_modified_at:
-            self.delete_queued_at = timezone.now()
-            self.save()
-            self.notify_changed(originating_user_id=originating_user_id)
-
-        delete_scratch_org_job.delay(self, originating_user_id=originating_user_id)
+        # Formerly a background job, currently synchronous:
+        self.delete()
```
Comment (jgerigmeyer) [inline metecho/api/models.py:882]: This method already internally calls `self.finalize_delete`

**Comment by jgerigmeyer** [comment]
@davisagli Do you have thoughts on my question about the UX flow for re-assigning a task role (with an org) from one user to another?

> I have some UX questions around this new flow. Right now, we show a warning before re-assigning if the org has unretrieved changes (because the org might be deleted, and we don't want to unintentionally lose those changes). Do we want to continue doing that always? Or first check if the org can be re-assigned and only show the warning if it can't be re-assigned? Or do away with the warning entirely?

File: metecho/api/models.py (hunk 9 above)
Comment (jgerigmeyer) [reply metecho/api/models.py:1047]: Actually, maybe not. But this brings up a related issue -- the front end does not subscribe to WS events for deleted objects. So it won't receive this message about un-deletion... We can talk through the options there.

Comment (wlonk) [reply metecho/api/models.py:1047]: What sort of options are you considering? The obvious one to me is to stay subscribed to deleted ScratchOrgs.

Comment (jgerigmeyer) [reply metecho/api/models.py:1047]: Right, but this may involve an org that was soft-deleted prior to a page refresh, so the front end never knew about it in the first place. It seems a bit much for the client to subscribe to every org every time, deleted or not?

I was thinking a new event (object un-deleted or whatnot), either on a new list-specific channel (a la https://trello.com/c/1TS05t3A) or on the User channel or something.

File: metecho/api/models.py (hunk 8 above)
Comment (jgerigmeyer) [reply metecho/api/models.py:883]: My understanding, per the Trello card, is that this new logic is used when _reassigning_ from one user to another ("When reassigning an org..."), or when _removing_ an assigned user ("When removing a user who has an org...").

If the user clicks "Delete Org", I understood that we would still actually delete the org on SF, and it would not be a candidate for re-use. Otherwise, as a user how do I clean up my orgs if I hit my SF scratch org limit? I don't think we want users to have to go off to SF to do that. I think this feature is all about how to handle org _auto_-deletion differently.

Comment (wlonk) [reply metecho/api/models.py:883]: Ah, I follow. To explain my understanding, I have to make explicit the distinction between local orgs and SF orgs. When a local org has no user assigned to it, we locally soft-delete it and don't touch the SF org. If we then have a use reassigned to the local org, we can reassign the attached SF org. If we explicitly delete a local org, we also bring down the SF org. This is clear..

(So I need to keep the queue-delete logic, for that case!)

Comment (wlonk) [reply metecho/api/models.py]: I think maybe I have implemented this? I am struggling to feel certain.

Comment (jgerigmeyer) [reply metecho/api/models.py]: @wlonk I think so! We still need to figure out a good solution to https://github.com/oddbird/Metecho/pull/173#discussion_r442273124

**Comment by wlonk** [comment]
[Optimize websocket subscriptions with bulk-subscription API.](https://trello.com/c/1TS05t3A/126-optimize-websocket-subscriptions-with-bulk-subscription-api)

**Comment by wlonk** [comment]
[Optimize websocket subscriptions with bulk-subscription API.](https://trello.com/c/1TS05t3A/126-optimize-websocket-subscriptions-with-bulk-subscription-api)

**Comment by wlonk** [comment]
@jgerigmeyer let me know if this list-subscribe works for you.

File: metecho/api/model_mixins.py [hunk 11]
```diff
@@ -82,6 +88,13 @@ def notify_changed(self, *, type_=None, originating_user_id, message=None):
             type_ or self.push_update_type, prepared_message,
         )
 
+    def notify_list_changed(self, *, type_, originating_user_id, message=None):
+        prepared_message = {"originating_user_id": originating_user_id}
+        prepared_message.update(message or {})
+        async_to_sync(push.push_message_about_instance)(
+            self, {"type": type_, "payload": prepared_message}
+        )
```
Comment (jgerigmeyer) [inline metecho/api/model_mixins.py:96]: @wlonk Any reason to do this instead of using `self._push_message`?

File: metecho/api/models.py [hunk 12]
```diff
@@ -1047,7 +1058,9 @@ def queue_refresh_org(self, *, originating_user_id):
         self.has_been_visited = False
         self.currently_refreshing_org = True
         self.save()
-        self.notify_changed(originating_user_id=originating_user_id)
+        self.notify_changed(
+            type_="SCRATCH_ORG_RECREATE", originating_user_id=originating_user_id,
+        )
```
Comment (jgerigmeyer) [inline metecho/api/models.py:1063]: @wlonk I don't think we want this to change -- it seems unrelated to the rest of this PR. From the perspective of the front end, "refreshing" an org is not the same as undeleting an org.

File: metecho/api/models.py [hunk 13]
```diff
@@ -1066,6 +1079,33 @@ def finalize_refresh_org(self, *, error=None, originating_user_id):
...
 
+    def queue_reassign(self, *, new_user, originating_user_id):
+        from .jobs import user_reassign_job
+
+        self.currently_refreshing_org = True
+        self.deleted_at = None
+        self.save()
+        self.notify_list_changed(
+            type_="SCRATCH_ORG_RECREATE", originating_user_id=originating_user_id,
+        )
```
Comment (jgerigmeyer) [inline metecho/api/models.py:1090]: @wlonk I'm wondering if we want to only fire this event if the org was previously `deleted_at != None`? Otherwise the org is just being re-assigned, not re-created.

File: metecho/api/models.py [hunk 14]
```diff
@@ -1066,6 +1079,33 @@ def finalize_refresh_org(self, *, error=None, originating_user_id):
...
+        if error is None:
+            self.notify_changed(
+                type_="SCRATCH_ORG_REASSIGN", originating_user_id=originating_user_id
+            )
+        else:
+            self.notify_scratch_org_error(
+                error=error,
+                type_="SCRATCH_ORG_REASSIGN_FAILED",
+                originating_user_id=originating_user_id,
+            )
```
Comment (jgerigmeyer) [inline metecho/api/models.py:1107]: @wlonk I'm trying to think through what we should do on the error case here... Probably soft-delete the org?

File: metecho/api/push.py [hunk 15]
```diff
@@ -63,6 +64,20 @@ async def push_message_about_instance(instance, message):
...
 
 
+async def push_message_about_list(cls, message):
+    model_name = cls._meta.model_name
+    group_name = CHANNELS_GROUP_NAME.format(model=model_name, id="list")
+    channel_layer = get_channel_layer()
+
+    new_message = deepcopy(message)
+    new_message["model_name"] = model_name
+    new_message["id"] = "list"
```
Comment (jgerigmeyer) [inline metecho/api/push.py:74]: @wlonk Should `"list"` move into `constants.py`, since it's used in multiple files?

File: metecho/consumers.py [hunk 16]
```diff
@@ -59,16 +59,29 @@ async def hydrate_message(self, content):
...
+                else:
+                    content["payload"]["model"] = await database_sync_to_async(
+                        instance.get_serialized_representation
+                    )(self.scope["user"])
+            # Handle list events:
             else:
+                class_ = self.get_class(model=model_name)
                 content["payload"]["model"] = await database_sync_to_async(
-                    instance.get_serialized_representation
+                    class_.get_list_serialized_representation
```
Comment (jgerigmeyer) [inline metecho/consumers.py:76]: @wlonk I understand why this makes theoretical sense from the perspective of the list channel... but practically I'm trying to think through how to handle all this data that the front end doesn't actually need. All we _really_ need in the new `SCRATCH_ORG_RECREATE` event is the one org instance that has been re-created, even if the event is sent on the list channel. Getting the full list of orgs isn't necessary, and presents a problem for the front end to determine _which_ org was re-created. (The client never even queries or knows about the _full_ list of orgs -- it only fetches them per-task, on-demand.)

Even if we leave some of this logic in place, could we also allow sending custom events about one instance on the list channel?

File: metecho/api/serializers.py [hunk 17]
```diff
@@ -364,23 +364,55 @@ def update(self, instance, validated_data):
...
+                    "dev", org, validated_data["assigned_dev"]
+                )
+                valid_commit = org.latest_commit == (
+                    instance.commits[0] if instance.commits else instance.origin_sha
+                )
+                if new_user and valid_commit and not reassigned_org:
+                    org.queue_reassign(
+                        new_user=new_user, originating_user_id=originating_user_id
+                    )
+                    reassigned_org = True
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:381]: @wlonk Note from today's mtg: We want to add an endpoint solely for the purpose of knowing whether an existing (dev) org can be reassigned from one user to another.

Comment (wlonk) [reply metecho/api/serializers.py:381]: This should just be an endpoint that does the same logic as is currently in `TaskSerializer._valid_reassign`, right? Or is there another part to it that I'm forgetting (cc @davisagli )

File: metecho/consumers.py (hunk 16 above)
Comment (wlonk) [reply metecho/consumers.py:76]: I was able to restructure this so you can subscribe to `list` and we can mark certain events as "for the list", but the individual object will still be serialized and sent. I think this should work.

I agree, this was theoretically pure but practically bad.

File: metecho/api/serializers.py (hunk 17 above)
Comment (jgerigmeyer) [reply metecho/api/serializers.py:381]: @wlonk I think that's correct, yes. The front end provides an org and a GH user, and the endpoint returns a boolean (effectively).

File: metecho/api/views.py [hunk 18]
```diff
@@ -228,6 +230,38 @@ def review(self, request, pk=None):
...
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        task = self.get_object()
+        role = serializer.validated_data["role"]
+        role_org_type = {
+            "assigned_qa": SCRATCH_ORG_TYPES.QA,
+            "assigned_dev": SCRATCH_ORG_TYPES.Dev,
+        }.get(role, None)
+        gh_uid = serializer.validated_data["gh_uid"]
+        # We want to consider soft-deleted orgs, too:
+        org = task.scratchorg_set.all().filter(org_type=role_org_type).first()
```
Comment (jgerigmeyer) [inline metecho/api/views.py:246]: @wlonk In this case don't we specifically only care about active orgs? That's why we can do `.first()` instead of iterating over all orgs -- this endpoint is only "can the _current active org_ for this task-role be reassigned to this new user?"

File: metecho/api/serializers.py [hunk 19]
```diff
@@ -364,23 +364,55 @@ def update(self, instance, validated_data):
         if instance.assigned_dev != validated_data["assigned_dev"]:
             if validated_data.get("should_alert_dev"):
                 self.try_send_assignment_emails(instance, "dev", validated_data, user)
-            orgs = instance.scratchorg_set.active().filter(
-                org_type=SCRATCH_ORG_TYPES.Dev
-            )
+            # We want to consider soft-deleted orgs, too:
+            orgs = instance.scratchorg_set.all().filter(org_type=SCRATCH_ORG_TYPES.Dev)
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:368]: @wlonk Is this iterating order too arbitrary? It seems possible that multiple orgs (some soft-deleted) could be potentially re-assigned... If there is an active org already in this task-role, we should definitely be checking that first -- then looking through soft-deleted orgs, probably in `-created_at` order?

File: metecho/api/views.py [hunk 20]
```diff
@@ -228,6 +230,38 @@ def review(self, request, pk=None):
...
+            None,
+        )
+        valid_commit = org and org.latest_commit == (
+            task.commits[0] if task.commits else task.origin_sha
+        )
+        return Response(
+            {
+                "can_reassign": bool(
+                    new_user
+                    and org.owner_sf_username == new_user.sf_username
```
Comment (jgerigmeyer) [inline metecho/api/views.py:259]: @wlonk Will this error if `org` is missing?

File: metecho/api/serializers.py [hunk 21]
```diff
@@ -361,26 +361,53 @@ def create(self, validated_data):
...
+                valid_commit = org.latest_commit == (
+                    instance.commits[0] if instance.commits else instance.origin_sha
+                )
+                if new_user and valid_commit and not reassigned_org:
+                    org.queue_reassign(
+                        new_user=new_user, originating_user_id=originating_user_id
+                    )
+                    reassigned_org = True
+                else:
+                    org.delete(originating_user_id=originating_user_id)
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:401]: @wlonk Right now this is also deleting the org on SF (via `model_mixins`, which calls `queue_delete`), which means it will not be available to be reused later. I'm not sure when we _do_ want to delete a scratch org on SF, vs when we do _not_ want to do that. I think we maybe want to distinguish between an org being explicitly deleted (because the user said "delete this org") vs deleted as a side-effect. I *think* we could do that by removing the logic in `model_mixins` that automatically deletes scratch orgs on SF when they're soft-deleted in Metecho (because in `perform_destroy` we're directly calling `queue_delete` to delete it on SF)... but I'd want a second set of eyes on that.

File: metecho/api/serializers.py [hunk 22]
```diff
@@ -361,26 +361,53 @@ def create(self, validated_data):
...
         validated_data.pop("should_alert_qa", None)
         return super().update(instance, validated_data)
 
+    def _handle_reassign(
+        self, type_, instance, validated_data, user, originating_user_id
+    ):
+        if (
+            getattr(instance, f"assigned_{type_}")
+            != validated_data[f"assigned_{type_}"]
+        ):
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:378]: @wlonk We're going through a lot of unnecessary logic here in the case where the assigned user has been removed. In that case I think we could bail early and just soft-delete the org.

File: metecho/api/serializers.py (hunk 21 above)
Comment (wlonk) [reply metecho/api/serializers.py:401]: I'm so confused by the intended behavior on this, at this point. I have no idea anymore. Maybe we should talk about it more?

Comment (jgerigmeyer) [reply metecho/api/serializers.py:401]: @wlonk I don't think anything has changed since our earlier discussion (https://github.com/oddbird/Metecho/pull/173#discussion_r442368141). You already made this change, so that it's now calling `org.delete()` instead of `org.queue_delete` -- with the intention of *not* deleting the org on SF automatically when the task role is reassigned. I'm pointing out that it's not quite working as intended yet, because the soft-delete eventually ends up calling `scratch_org.queue_delete()` anyway, as part of the soft-deletion.

In fact, that's already a bug in the current codebase, even before this PR: When you DELETE an org, it triggers `queue_delete`, which then calls `delete`, which then calls `queue_delete` _again_.

Comment (wlonk) [reply metecho/api/serializers.py:401]: `ScratchOrg.delete` doesn't call `queue_delete` though? We need to talk about this with code in front of us.

Comment (jgerigmeyer) [reply metecho/api/serializers.py:401]: @wlonk It does, via https://github.com/oddbird/Metecho/blob/main/metecho/api/model_mixins.py#L186 and https://github.com/oddbird/Metecho/blob/main/metecho/api/model_mixins.py#L226

Would you prefer to jump on a call? I'm around.

Comment (wlonk) [reply metecho/api/serializers.py:401]: What kind of terrible design queues a delete inside a `notify_`! This is why I need more code review; I think that's terrible, and I wrote it.

File: metecho/api/serializers.py [hunk 23]
```diff
@@ -361,26 +361,53 @@ def create(self, validated_data):
...
+                *instance.scratchorg_set.inactive().filter(org_type=org_type),
+            ]
+            for org in orgs:
+                new_user = self._valid_reassign(
+                    type_, org, validated_data[f"assigned_{type_}"]
+                )
+                valid_commit = org.latest_commit == (
+                    instance.commits[0] if instance.commits else instance.origin_sha
+                )
+                if new_user and valid_commit and not reassigned_org:
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metecho/api/serializers.py:395]: Hm, I actually have no idea how to check whether the org has been deleted on Salesforce. And we _have_ to do that, because it might be deleted manually early without our knowledge, before the expiration date.

[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply metecho/api/serializers.py:395]: @wlonk Is it essentially what we're doing in `sf_run_flow` `refresh_access_token` -- catching a `HTTPError`?

[the remark this answers is not in the document]
Comment (wlonk) [reply metecho/api/serializers.py:395]: Probable.

File: metecho/api/model_mixins.py [hunk 24]
```diff
@@ -226,15 +217,7 @@ class Meta:
...
-        if self.__class__.__name__ == "ScratchOrg":
-            try:
-                self.queue_delete(originating_user_id=None)
-            except Exception:  # pragma: nocover
-                # If there's a problem deleting it, it's probably
-                # already been deleted.
-                pass
-        else:
-            self.notify_changed(type_="SOFT_DELETE", originating_user_id=None)
+        self.notify_changed(type_="SOFT_DELETE", originating_user_id=None)
```
[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply metecho/api/model_mixins.py:220]: Put differently, I think the only time we _don't_ want to also delete orgs on SF is when we're re-assigning or removing a task role.

File: metecho/api/serializers.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply metecho/api/serializers.py]: Looks good!

File: metecho/api/model_mixins.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply metecho/api/model_mixins.py:220]: @wlonk Tested it out; this is working as expected. 🎉 I think resolving this comment is the last piece on this PR?

**Comment by jgerigmeyer** [comment]
[Rename "ms_" and "ms-" items to "metecho_" and "metecho-".](https://trello.com/c/QBB14QQV/212-rename-ms-and-ms-items-to-metecho-and-metecho)
